Implements specialized subsetOf for HashSet

rklaehn opened the pull request:
Fixes SI-7326. This also adds a basic test for subsetOf that was missing before.

Review by @Ichoran, @soc

@Ichoran: feel free to integrate the subsetOf test into your collection comparison test suite and remove the testCorrectness part of t7326.scala

rklaehn commented:
Here is a small benchmark that I used to compare the old with the new implementation, and the results on my machine (haswell i7 notebook CPU). As you can see, time ratio is always >1 and often significantly so. 

If anybody asks why he should care about subsetOf: it is a fundamental set operation and it is used by equality.

---

import ichi.bench.Thyme
import scala.collection.immutable2.{HashSet => HashSet2}
import scala.collection.immutable.HashSet

object HashSetSubsetOfBenchmark extends App {

  def timeToString(time:Double) = (time*1000000).formatted("%.3f")

  val th = Thyme.warmed()

  def subsetComparison[T](sizes:Seq[Int], empty0:Set[T], empty1:Set[T], mkKey:Int => T) {

```
val empty0 = HashSet.empty
val o = sizes.max + 1
for {
  i <- sizes
  k <- sizes
  if i >= k
}
{
  val ikeys = (0 to i).map(mkKey)
  val kkeys = (0 to k).map(mkKey)
  val a0 = empty0 ++ ikeys
  val b0 = empty0 ++ kkeys
  val a1 = empty1 ++ ikeys
  val b1 = empty1 ++ kkeys
  // val c = b + o
  // manual warmup
  for(i<-0 until 100) {
    a0.subsetOf(b0)
    a1.subsetOf(b1)
  }
  val r = th.pbenchOff(title = "subsetOf")(a1.subsetOf(b1))(a0.subsetOf(b0))
  //val (r,t) = th.timePair(b0.subsetOf(a0))
  //val runtime = timeToString(t.runTime)

  //println(s"$i\t$k\t$runtime")
}
```

  }

  val sizes = Seq(0,1,10,100,1000,10000)
  subsetComparison(sizes, HashSet.empty[Int], HashSet2.empty[Int], identity)
}

---

Benchmark comparison (in 424.0 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    2.57225   95% CI 2.40030 - 2.74420   (n=225)
    First     5.620 ns   95% CI 5.269 ns - 5.970 ns
    Second    14.45 ns   95% CI 14.10 ns - 14.81 ns
  Individual benchmarks not fully consistent with head-to-head (p ~= 2.145e-11)
    First     7.059 ns   95% CI 6.833 ns - 7.286 ns
    Second    14.29 ns   95% CI 14.06 ns - 14.52 ns
Benchmark comparison (in 403.6 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    5.29031   95% CI 4.73939 - 5.84122   (n=150)
    First     5.486 ns   95% CI 4.924 ns - 6.047 ns
    Second    29.02 ns   95% CI 28.46 ns - 29.58 ns
  Individual benchmarks not fully consistent with head-to-head (p ~= 0.0023)
    First     5.822 ns   95% CI 5.682 ns - 5.962 ns
    Second    26.68 ns   95% CI 26.06 ns - 27.30 ns
Warning: order-of-evaluation effects in timing (p ~= 0.0162) of up to 75.598%
Benchmark comparison (in 225.4 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    1.58663   95% CI 1.49723 - 1.67602   (n=45)
    First     24.92 ns   95% CI 23.73 ns - 26.11 ns
    Second    39.54 ns   95% CI 38.35 ns - 40.72 ns
Benchmark comparison (in 256.0 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    1.88941   95% CI 1.76823 - 2.01058   (n=100)
    First     17.42 ns   95% CI 16.43 ns - 18.41 ns
    Second    32.92 ns   95% CI 31.93 ns - 33.90 ns
  Individual benchmarks not fully consistent with head-to-head (p ~= 4.778e-13)
    First     7.261 ns   95% CI 7.013 ns - 7.509 ns
    Second    23.99 ns   95% CI 23.74 ns - 24.24 ns
Benchmark comparison (in 277.8 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    4.41080   95% CI 3.90919 - 4.91242   (n=100)
    First     7.567 ns   95% CI 6.728 ns - 8.407 ns
    Second    33.38 ns   95% CI 32.54 ns - 34.22 ns
Benchmark comparison (in 181.2 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    1.46075   95% CI 1.40838 - 1.51311   (n=20)
    First     99.14 ns   95% CI 96.21 ns - 102.1 ns
    Second    144.8 ns   95% CI 141.9 ns - 147.8 ns
Benchmark comparison (in 362.6 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    4.04342   95% CI 3.79419 - 4.29265   (n=150)
    First     7.041 ns   95% CI 6.620 ns - 7.463 ns
    Second    28.47 ns   95% CI 28.05 ns - 28.89 ns
Benchmark comparison (in 362.0 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    4.10171   95% CI 3.85858 - 4.34484   (n=150)
    First     7.611 ns   95% CI 7.173 ns - 8.049 ns
    Second    31.22 ns   95% CI 30.78 ns - 31.66 ns
Benchmark comparison (in 364.3 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    4.25556   95% CI 4.00209 - 4.50903   (n=150)
    First     7.351 ns   95% CI 6.924 ns - 7.777 ns
    Second    31.28 ns   95% CI 30.85 ns - 31.71 ns
  Individual benchmarks not fully consistent with head-to-head (p ~= 0.0015)
    First     7.855 ns   95% CI 7.711 ns - 8.000 ns
    Second    30.65 ns   95% CI 30.49 ns - 30.81 ns
Benchmark comparison (in 807.0 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    1.86166   95% CI 1.84596 - 1.87737   (n=20)
    First     1.122 us   95% CI 1.116 us - 1.129 us
    Second    2.089 us   95% CI 2.076 us - 2.102 us
Benchmark comparison (in 205.6 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    5.92433   95% CI 5.07447 - 6.77419   (n=67)
    First     6.776 ns   95% CI 5.817 ns - 7.734 ns
    Second    40.14 ns   95% CI 39.18 ns - 41.10 ns
Benchmark comparison (in 150.6 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    6.08663   95% CI 5.20255 - 6.97070   (n=45)
    First     7.237 ns   95% CI 6.200 ns - 8.274 ns
    Second    44.05 ns   95% CI 43.01 ns - 45.09 ns
Benchmark comparison (in 154.3 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    6.10689   95% CI 5.20693 - 7.00685   (n=45)
    First     7.407 ns   95% CI 6.330 ns - 8.484 ns
    Second    45.23 ns   95% CI 44.15 ns - 46.31 ns
Benchmark comparison (in 253.9 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    6.55215   95% CI 5.71725 - 7.38706   (n=67)
    First     7.646 ns   95% CI 6.683 ns - 8.609 ns
    Second    50.10 ns   95% CI 49.14 ns - 51.06 ns
Benchmark comparison (in 697.0 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    2.34001   95% CI 2.27947 - 2.40056   (n=20)
    First     19.21 us   95% CI 18.95 us - 19.46 us
    Second    44.94 us   95% CI 43.94 us - 45.94 us
Benchmark comparison (in 276.7 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    5.51851   95% CI 4.84467 - 6.19235   (n=100)
    First     6.901 ns   95% CI 6.072 ns - 7.730 ns
    Second    38.08 ns   95% CI 37.26 ns - 38.91 ns
Benchmark comparison (in 152.8 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    6.48996   95% CI 5.52179 - 7.45814   (n=45)
    First     6.889 ns   95% CI 5.873 ns - 7.904 ns
    Second    44.71 ns   95% CI 43.69 ns - 45.72 ns
Benchmark comparison (in 155.1 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    6.63951   95% CI 5.61612 - 7.66289   (n=45)
    First     6.960 ns   95% CI 5.899 ns - 8.021 ns
    Second    46.21 ns   95% CI 45.15 ns - 47.27 ns
Benchmark comparison (in 165.5 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    7.20100   95% CI 6.11874 - 8.28327   (n=45)
    First     6.940 ns   95% CI 5.907 ns - 7.973 ns
    Second    49.97 ns   95% CI 48.94 ns - 51.01 ns
Benchmark comparison (in 172.1 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    6.08398   95% CI 5.28588 - 6.88208   (n=45)
    First     8.294 ns   95% CI 7.220 ns - 9.367 ns
    Second    50.46 ns   95% CI 49.39 ns - 51.53 ns
Benchmark comparison (in 848.4 ms): subsetOf
Significantly different (p ~= 0)
  Time ratio:    2.22861   95% CI 2.11404 - 2.34318   (n=20)
    First     300.0 us   95% CI 288.5 us - 311.5 us
    Second    668.6 us   95% CI 645.6 us - 691.6 us
Warning: order-of-evaluation effects in timing (p ~= 0.0011) of up to 27.680%

soc commented:
Impressive performance improvements, by the way! :-)

rklaehn commented:
Thanks. There are a lot of such improvements to be had all over the collections library by overriding the builder-based implementations. Although I am sometimes surprised how well the builder-based ones perform.

What's the best way to make subsetOf(that:HashSet[A]) private as discussed? Just push a new update, or close and redo the pull request?

soc commented:
Wow, the improvements look great! I'd not bother much about creating new PR's. I woudl just squash the first three commits into one and have the ScalaDoc additions in a second commit.

LGTM!

rklaehn commented:
OK, there are now just two commits. One for the specialized subsetOf and the corresponding test case, the other for the updated documentation.

The documentation would have to be more or less repeated for HashMap, since that uses the exact same Trie structure internally. It really is a shame that we can not unify HashSet and HashMap due to efficiency reasons.

Ichoran commented:
LGTM!  Thanks!

rklaehn commented:
@soc: you want me to update the documentation of HashSet in a similar way to the second commit, or is it enough to have this in HashSet?

Ichoran commented:
HashMap should just reference the description in HashSet.  (And don't add that to this pull request.)

rklaehn commented:
OK, makes sense. I will add the reference in a separate documentation PR.
But I guess it does not make sense to add the reference until this PR is
merged.

On Sun, Dec 29, 2013 at 6:00 PM, Ichoran notifications@github.com wrote:

> HashMap should just reference the description in HashSet. (And don't add
> that to this pull request.)
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/3315#issuecomment-31320490
> .

soc commented:
@rklaehn I think both classes should get the ScalaDoc. Maybe you could use ScalaDoc's macro facility to keep the documentation in one place mostly?

rklaehn commented:
I am not familiar with the macro facility, but I will take a look. But
looking at HashMap, it will need some additional documentation anyway. For
example the fact that both key, value and kv are stored in a HashMap1.

Has anybody actually verified that caching the kv is a good idea? From all
I know about the JVM, short-lived objects are basically free whereas
longer-lived objects are not.

On Sun, Dec 29, 2013 at 6:30 PM, soc notifications@github.com wrote:

> @rklaehn https://github.com/rklaehn I think both classes should get the
> ScalaDoc. Maybe you could use ScalaDoc's macro facility to keep the
> documentation in one place mostly?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/3315#issuecomment-31321107
> .

Ichoran commented:
I doubt that the utility of storing kv has been checked lately.
  --Rex

On Sun, Dec 29, 2013 at 9:56 AM, Rüdiger Klaehn notifications@github.comwrote:

> I am not familiar with the macro facility, but I will take a look. But
> looking at HashMap, it will need some additional documentation anyway. For
> example the fact that both key, value and kv are stored in a HashMap1.
> 
> Has anybody actually verified that caching the kv is a good idea? From all
> I know about the JVM, short-lived objects are basically free whereas
> longer-lived objects are not.
> 
> On Sun, Dec 29, 2013 at 6:30 PM, soc notifications@github.com wrote:
> 
> > @rklaehn https://github.com/rklaehn I think both classes should get
> > the
> > ScalaDoc. Maybe you could use ScalaDoc's macro facility to keep the
> > documentation in one place mostly?
> > 
> > —
> > Reply to this email directly or view it on GitHub<
> > https://github.com/scala/scala/pull/3315#issuecomment-31321107>
> > .
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/3315#issuecomment-31321599
> .

rklaehn commented:
I opened an issue. https://issues.scala-lang.org/browse/SI-8108 .

Removing kv should be relatively easy. The question is how to test that
there is no performance degradation. I am also not 100% confident that
there is a watertight unit test for HashMap.

On Sun, Dec 29, 2013 at 7:53 PM, Ichoran notifications@github.com wrote:

> I doubt that the utility of storing kv has been checked lately.
> --Rex
> 
> On Sun, Dec 29, 2013 at 9:56 AM, Rüdiger Klaehn notifications@github.comwrote:
> 
> > I am not familiar with the macro facility, but I will take a look. But
> > looking at HashMap, it will need some additional documentation anyway.
> > For
> > example the fact that both key, value and kv are stored in a HashMap1.
> > 
> > Has anybody actually verified that caching the kv is a good idea? From
> > all
> > I know about the JVM, short-lived objects are basically free whereas
> > longer-lived objects are not.
> > 
> > On Sun, Dec 29, 2013 at 6:30 PM, soc notifications@github.com wrote:
> > 
> > > @rklaehn https://github.com/rklaehn I think both classes should get
> > > the
> > > ScalaDoc. Maybe you could use ScalaDoc's macro facility to keep the
> > > documentation in one place mostly?
> > > 
> > > —
> > > Reply to this email directly or view it on GitHub<
> > > https://github.com/scala/scala/pull/3315#issuecomment-31321107>
> > > .
> > 
> > —
> > Reply to this email directly or view it on GitHub<
> > https://github.com/scala/scala/pull/3315#issuecomment-31321599>
> > .
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/3315#issuecomment-31322652
> .

Ichoran commented:
Certainly isn't anything comprehensive, but

  https://github.com/scala/scala/pull/3306

is a reasonable start.

  --Rex

On Sun, Dec 29, 2013 at 11:06 AM, Rüdiger Klaehn
notifications@github.comwrote:

> I opened an issue. https://issues.scala-lang.org/browse/SI-8108 .
> 
> Removing kv should be relatively easy. The question is how to test that
> there is no performance degradation. I am also not 100% confident that
> there is a watertight unit test for HashMap.
> 
> On Sun, Dec 29, 2013 at 7:53 PM, Ichoran notifications@github.com
> wrote:
> 
> > I doubt that the utility of storing kv has been checked lately.
> > --Rex
> > 
> > On Sun, Dec 29, 2013 at 9:56 AM, Rüdiger Klaehn <
> > notifications@github.com>wrote:
> > 
> > > I am not familiar with the macro facility, but I will take a look. But
> > > looking at HashMap, it will need some additional documentation anyway.
> > > For
> > > example the fact that both key, value and kv are stored in a HashMap1.
> > > 
> > > Has anybody actually verified that caching the kv is a good idea? From
> > > all
> > > I know about the JVM, short-lived objects are basically free whereas
> > > longer-lived objects are not.
> > > 
> > > On Sun, Dec 29, 2013 at 6:30 PM, soc notifications@github.com
> > > wrote:
> > > 
> > > > @rklaehn https://github.com/rklaehn I think both classes should
> > > > get
> > > > the
> > > > ScalaDoc. Maybe you could use ScalaDoc's macro facility to keep the
> > > > documentation in one place mostly?
> > > > 
> > > > —
> > > > Reply to this email directly or view it on GitHub<
> > > > https://github.com/scala/scala/pull/3315#issuecomment-31321107>
> > > > .
> > > 
> > > —
> > > Reply to this email directly or view it on GitHub<
> > > https://github.com/scala/scala/pull/3315#issuecomment-31321599>
> > > .
> > 
> > —
> > Reply to this email directly or view it on GitHub<
> > https://github.com/scala/scala/pull/3315#issuecomment-31322652>
> > .
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/3315#issuecomment-31322897
> .

Ichoran commented on src/library/scala/collection/immutable/HashSet.scala:
```diff
@@ -12,9 +12,9 @@ package scala
 package collection
 package immutable
 
-import scala.annotation.unchecked.{ uncheckedVariance => uV }
```
I think this is going to conflict with union/diff merge that uses uV?  Or am I mistaken?

rklaehn replied:
I think it should be OK. However, I did not manage to make the big PRs completely independent of each other. So I will probably have to rebase them once this is committed.

rklaehn committed:
Implements specialized subsetOf for HashSet

Fixes SI-7326. This also adds a basic test for subsetOf that was missing before.

rklaehn committed:
Improved documentation of HashTrieSet internals

Added some more general documentation about how levels work and how the
bitmap field encodes the position of elems.

rklaehn commented:
I resubmitted the two commits to fix a cosmetic issue.
